### PR TITLE
Fixed issue #224: ion_int_from_long doesn't return

### PR DESCRIPTION
**Pull request by zslayton**
This PR addresses a corner case in `ion_int_from_long` that would cause an infinite loop when MIN_INT64 is passed in as the long value. A detailed description of the problem can be found in issue #224.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by tgregg** [comment]
The fix looks good, but can you please include the unit test?

**Comment by zslayton** [comment]
> The fix looks good, but can you please include the unit test?

Polishing up my unit test has surfaced another problem. 😞 See #226. I'm going to try and roll a fix for that into this PR.

File: ionc/ion_int.c [hunk 1]
```diff
@@ -1599,28 +1611,63 @@ SIZE _ion_int_abs_bytes_signed_length_helper(const ION_INT *iint)
...
     digits = iint->_digits;
     end    = digits + iint->_len;
-    while (digits < end) {
+
+    // iint's magnitude is stored in an array of 31-bit II_DIGIT values. Magnitudes that
+    // require 63 or 64 bits to represent will then take 3 II_DIGITS, only using 1 to 2
+    // bits from the most significant II_DIGIT.
+    const uint32_t max_digits_per_int64_t = 3;
+    const uint32_t whole_digits_per_int64_t = 2;
+    const uint32_t max_partial_digit_value = 3; // == 0b11, two populated bits
```
Comment (zslayton) [inline ionc/ion_int.c:1628]: These constants went through a few iterations before I settled on this.

I tried re-calculating these values based on `sizeof(int64_t)` and `II_BITS_PER_II_DIGIT`, but it was pretty verbose:

```c
const uint32_t bits_per_int64_t = sizeof(int64_t) * II_BITS_PER_BYTE;
const float digits_per_int64_t = (float) bits_per_int64_t / (float) II_BITS_PER_II_DIGIT;
const uint32_t max_digits = ceilf(digits_per_int64_t);
const uint32_t whole_digits = floorf(digits_per_int64_t);
const uint32_t max_bits_in_partial_digit = bits_per_int64_t % II_BITS_PER_II_DIGIT;
const uint32_t max_partial_digit_value = (1<<max_bits_in_partial_digit) - 1;
```

While a compiler would've probably been able to reduce these to zero-cost constants at build time, the complexity it introduced for folks trying to read the code felt too expensive.

Given that we're using a type name that will always be 64-bits (`int64_t` instead of, say, `long`), this would only ever need to be reevaluated if the definition of `II_BITS_PER_II_DIGIT` (currently `31`) ever changed, which won't happen any time soon. The code above didn't seem to carry its own weight. I opted to just use the resulting values instead.

I also tried defining macros for these values in `ion_int.h`, but it felt like overkill; readers would have to track down their definitions in another file despite this being the only place they're ever used.

File: ionc/ion_int.c [hunk 2]
```diff
@@ -625,25 +630,32 @@ iERR ion_int_from_long(ION_INT *iint, int64_t value)
...
     }
 
-    if ((is_neg = (value < 0)) == TRUE) {
-        value = -value;
+    is_negative = value < 0;
+    magnitude = (uint64_t) value;
+    if (is_negative) {
+        // This negates an unsigned value, which is well-defined behavior. Doing so handles the
+        // MIN_INT64 case: an int64_t whose absolute value is too large to be stored in an int64_t.
+        magnitude = -magnitude;
```
Comment (zslayton) [inline ionc/ion_int.c:638]: Here's [a helpful StackOverflow answer](https://stackoverflow.com/questions/17313579/is-there-a-safe-way-to-get-the-unsigned-absolute-value-of-a-signed-integer-with#17313717) that includes the relevant text of the C spec.

File: ionc/ion_int.c [hunk 3]
```diff
@@ -615,8 +616,12 @@ iERR ion_int_from_long(ION_INT *iint, int64_t value)
 {
     iENTER;
     SIZE  ii_length, digit_idx;
-    int64_t temp;
-    BOOL    is_neg;
+    // Stores the unsigned magnitude of the provided int64_t value. This variable must be
+    // unsigned to accommodate the absolute value of MIN_INT64, which requires 64 bits to store.
+    uint64_t magnitude;
+    // Used for shifting operations that consume the variable.
+    uint64_t temp_magnitude;
```
Comment (zslayton) [inline ionc/ion_int.c:623]: `magnitude` is bit-shifted to zero twice. Given that it takes a small amount of expense to calculate it the first time, I introduced this temporary variable that could be reset to the value of `magnitude` between bit-shifting operations.

File: test/test_ion_integer.cpp [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    // Initialize the Ion integer, setting its owner to NULL.
+    IONCHECK(ion_int_init(&iint, NULL));
+    // Populate the Ion integer with the provided int64_t value.
+    IONCHECK(ion_int_from_long(&iint, value_in));
+    // Read the Ion integer's value back out into the output int64_t.
+    IONCHECK(ion_int_to_int64(&iint, value_out));
+    iRETURN;
+}
+
+TEST(IonInteger, IIntToInt64RoundTrip) {
```
Comment (tgregg) [inline test/test_ion_integer.cpp:32]: Should we also add a test that verifies overflow is detected when expected?

**Review by tgregg — APPROVED** [review]

**Review by tgregg — APPROVED** [review]
